Add virtual-kubelet Grafana dashboard (#1807)
liqotech/liqo: docs/usage/prometheus-metrics.md

## Scraping metrics

> How metrics are scraped will depend on how you are operating your **Prometheus server(s)**.

Reviewer: ```suggestion
Metrics are scraped depending on how your **Prometheus server(s)** has been configured.
```

> Liqo presumes that the [Prometheus Operator](https://github.com/prometheus-operator/prometheus-operator) is being used to run Prometheus, providing a **ServiceMonitor** resource for each component and a **PodMonitor** resource for the Virtual kubelet.

Reviewer: ```suggestion
Liqo assumes that you leverage the [Prometheus Operator](https://github.com/prometheus-operator/prometheus-operator) is being used to run Prometheus, hence providing a **ServiceMonitor** resource for each component and a **PodMonitor** resource for the Virtual kubelet.
```

> Metrics are **disabled** by default.

Reviewer: ```suggestion
Gathering of Liqo metrics is **disabled** by default.
```

> In order to allow Prometheus to scrape metrics from the Liqo components, you should set the `--enable-metrics` *liqoctl* flag during installation (cf. [installation customization options](InstallCustomization)).

Reviewer: ```suggestion
To allow Prometheus to scrape metrics from Liqo, you should set the `--enable-metrics` *liqoctl* flag during installation (cf. [installation customization options](InstallCustomization)), which configures Liqo to expose metrics, and it creates the proper **ServiceMonitor** and **PodMonitor** resources.
```

> This flag enables the metrics exposition, the **PodMonitor** and the **ServiceMonitor** resources creation.

Reviewer: ```suggestion
```

## Scraping metrics

Metrics are scraped depending on how your **Prometheus server(s)** has been configured.

> Liqo assumes that you leverage the [Prometheus Operator](https://github.com/prometheus-operator/prometheus-operator) is being used to run Prometheus, hence providing a **ServiceMonitor** resource for each component and a **PodMonitor** resource for the Virtual kubelet.

Reviewer: ```suggestion
Liqo assumes that you leverage the [Prometheus Operator](https://github.com/prometheus-operator/prometheus-operator) to run Prometheus, hence providing a **ServiceMonitor** resource for each component and a **PodMonitor** resource for the Virtual kubelet.
```

As presented in the screenshot below, it includes an overview section presenting the overall cross-cluster throughput, followed by detailed per-peering throughput and latency information.

![Grafana Network Dashboard](/_static/images/usage/prometheus-metrics/network-dashboard.png)

## Virtual kubelet metrics

These metrics are available for each peered remote cluster, providing statistics about the reflected resources:

> - **liqo_virtual_kubelet_reflection_item_counter**: the number of reflected resources successfully processed.

Reviewer: Not clear what "resources" means. I would suggest to rephrase such as:
... the number of reflected resources (e.g., pods, ...)
(list, in the parenthesis, all the objects that are counted)

Author: Ok, I will rephrase that. For resources, I meant pods, configmaps, services and others.

> - **liqo_virtual_kubelet_reflection_item_counter**: the number of reflected resources successfully processed (e.g., Pod, ConfigMap, Secret, Service, ServiceAccount, EndpointSlice, Ingress and PersistentVolumeClaim).

Reviewer: ```suggestion
- **liqo_virtual_kubelet_reflection_item_counter**: the number of resources that are currently successfully reflected (e.g., Pod, ConfigMap, Secret, Service, ServiceAccount, EndpointSlice, Ingress and PersistentVolumeClaim). This number can increase/decrease over time, and it may reach zero when two peered clusters have no reflected resources.
```

> - **liqo_virtual_kubelet_reflection_error_counter**: the number of transient errors during the reflection phase.

Reviewer: I would suggest to make an example here as well.
E.g.:
.... during the reflection phase, e.g., due to ....
(and list 2-3 reasons we may have errors here)

Author: Ok, I will add some examples.

Author: While discussing it with @aleoli, the main motivation behind the errors is that the issue is caused by temporary race conditions that can be resolved by retrying the synchronization. These conditions mainly occur when some of the requested resources are not yet fully configured. There could be three specific cases to discuss (e.g., reflector not completely initialized or fallback not completely initialized), but we think it would be too detailed to add to the documentation.

Reviewer: @aleoli @damianot98 Reporting an error is something that *should* be interesting for the user. For me, as a customer, knowing that are errors and (a) knowing nothing about why they are here, (b) having no way to solve them, means that this counter is completely useless. Even worse, it makes me crazy, as I know that something is not working, and I cannot do anything to solve the problem.
Please consider this and improve the documentation according to the above points.

Author: @frisso I extended the previous example, providing the three main reasons why errors occur in the reflection stage. Although it is highly specific, it would be more helpful from a user point of view.

> - **liqo_virtual_kubelet_reflection_error_counter**: the number of transient errors during the reflection phase (e.g., due to  temporary race conditions that can be resolved by retrying the synchronization, these conditions mainly occur when some of the requested resources are not yet fully configured).

Reviewer: ```suggestion
- **liqo_virtual_kubelet_reflection_error_counter**: the number of transient errors during the reflection phase (e.g., due to temporary race conditions that can be resolved by retrying the synchronization; these conditions mainly occur when some of the requested resources are not yet fully configured).
```

### Grafana dashboard

> We offer a {download}`sample Grafana dashboard </_downloads/grafana/liqovirtualkubelet.json>` that allows you to monitor the reflected resources for every componenet of the virtual-kubelet. As shown in the screenshot below, it contains visual representations of the total number of reflected resources and the average rate per second. Additionally, there are detailed tables that provide information on the total number of resources of each type, as well as an overall summary of all reflected items during a certain time period.

Reviewer: ```suggestion
We offer a {download}`sample Grafana dashboard </_downloads/grafana/liqovirtualkubelet.json>` that allows you to monitor the reflected resources for each component of the virtual-kubelet.
As shown in the screenshot below, it contains visual representations of the total number of reflected resources and the average rate per second.
Additionally, there are detailed tables that provide information on the total number of each type of resource, as well as an overall summary of all reflected items during a certain time period.
```
